### PR TITLE
[Lang] Support __len__ and __iter__ on ti.Matrix

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1188

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

Before this, the following code will result in a confusing error:
```py
import taichi as ti

a = ti.Vector.field(2, float, ())

a[None] = [2, 3]             # OK
a[None] = ti.Vector([2, 3])  # ERROR
```
```
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-51ftg4mj
[Taichi] <dev mode>, llvm 10.0.0, commit 41708c5e, linux, python 3.8.3
[Taichi] materializing...
Traceback (most recent call last):
  File "a.py", line 6, in <module>
    a[None] = ti.Vector([2, 3])  # ERROR
  File "/home/bate/Develop/taichi/python/taichi/lang/util.py", line 212, in wrapped
    return func(*args, **kwargs)
  File "/home/bate/Develop/taichi/python/taichi/lang/matrix.py", line 400, in __setitem__
    item = [[i] for i in item]
  File "/home/bate/Develop/taichi/python/taichi/lang/matrix.py", line 400, in <listcomp>
    item = [[i] for i in item]
  File "/home/bate/Develop/taichi/python/taichi/lang/util.py", line 212, in wrapped
    return func(*args, **kwargs)
  File "/home/bate/Develop/taichi/python/taichi/lang/matrix.py", line 394, in __getitem__
    return self(i, j)
  File "/home/bate/Develop/taichi/python/taichi/lang/matrix.py", line 229, in __call__
    return self.entries[self.linearize_entry_id(*args)]
  File "/home/bate/Develop/taichi/python/taichi/lang/matrix.py", line 220, in linearize_entry_id
    assert 0 <= args[0] < self.n, \
AssertionError: The 0-th matrix index is out of range: 0 <= 2 < 2
```
Why can't we assign `ti.Vector` to vector field elements?

It seems to be `list(v)` is accessing out of bound, so I fix this by adding a custom `__iter__` and `__len__`, that returns the vector entries.

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -410,6 +412,15 @@ def __setitem__(self, indices, item):
         j = 0 if len(indices) == 1 else indices[1]
         self.set_entry(i, j, item)
 
+    def __len__(self):
+        return self.n
```
Comment (k-ye) [inline python/taichi/lang/matrix.py:416]: What if `self.m > 1`? Shall we use `self.m * self.n`?

File: tests/python/test_matrix.py [hunk 2]
```diff
@@ -188,7 +188,31 @@ def func():
...
-@ti.all_archs
+@ti.test(arch=ti.cpu)
+def test_vector_to_list():
+    a = ti.Vector.field(2, float, ())
+
+    data = [2, 3]
+    b = ti.Vector(data)
+    assert list(b) == data
+
+    a[None] = b
```
Comment (k-ye) [inline tests/python/test_matrix.py:199]: nit: also assert `len(a) == 2`?

**Review by k-ye — APPROVED** [review]
Thanks! I left a few nits

File: python/taichi/lang/matrix.py (hunk 1 above)
Comment (archibate) [reply python/taichi/lang/matrix.py:416]: Please see `__iter__`, we return `[[2, 3, 4], [4, 5, 6]]` for `list(a)`, so `len(list(a))` is actually `a.n` in this case, let's make `len(a) == len(list(a))`.

File: tests/python/test_matrix.py [hunk 3]
```diff
@@ -188,7 +188,31 @@ def func():
...
 
 
-@ti.all_archs
+@ti.test(arch=ti.cpu)
+def test_vector_to_list():
+    a = ti.Vector.field(2, float, ())
+
+    data = [2, 3]
+    b = ti.Vector(data)
+    assert list(b) == data
```
Comment (archibate) [inline tests/python/test_matrix.py:197]:
```suggestion
    assert list(b) == data
    assert len(b) == data
```

File: tests/python/test_matrix.py [hunk 4]
```diff
@@ -188,7 +188,31 @@ def func():
...
+    assert all(a[None].value == ti.Vector(data))
+
+
+@ti.test(arch=ti.cpu)
+def test_matrix_to_list():
+    a = ti.Matrix.field(2, 3, float, ())
+
+    data = [[2, 3, 4], [5, 6, 7]]
+    b = ti.Matrix(data)
+    assert list(b) == data
```
Comment (archibate) [inline tests/python/test_matrix.py:209]:
```suggestion
    assert list(b) == data
    assert len(b) == len(data)
```

**Pull request merged**